### PR TITLE
Dev tls cipher suite v4

### PR DESCRIPTION
**Pull request by fooinha**
- Allows tls app-layer rules to match TLS handshake cypher suites
Example rules:

```
alert tls any any -> any any
(msg:"SURICATA TLS CIPHER SUITE CLIENT"; 
tls_cipher_suite:client:TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,0xcca9;
sid:66666666; 
rev:1;
)
```
```
alert tls any any -> any any 
(msg:"SURICATA TLS CIPHER SUITE SERVER";
tls_cipher_suite:server:0xcca8;
sid:66666667;
rev:1;
)
```

- Custom tls logging for server cypher suite %C 

- Documentation

Replaces #2503.

fyi, @thus

### CODE REVIEW

File: src/detect-tls-cipher-suite.c [hunk 1]
```diff
@@ -0,0 +1,1093 @@
...
+    {0xCCAC, "TLS_ECDHE_PSK_WITH_CHACHA20_POLY1305_SHA256"},
+    {0xCCAD, "TLS_DHE_PSK_WITH_CHACHA20_POLY1305_SHA256"},
+    {0xCCAE, "TLS_RSA_PSK_WITH_CHACHA20_POLY1305_SHA256"},
+    {0xFFFF, "" }
+
+};
+
+typedef struct DetectTlsCipherSuiteData_ {
+    uint32_t direction;         /**< STREAM_TOSERVER or STREAM_TOCLIENT */
+    uint32_t length;            /** < Number of cipher suites */
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: in SSLState the num_cipher_suites is u16, why do we have u32 here?

File: src/detect-tls-cipher-suite.c [hunk 2]
```diff
@@ -0,0 +1,1093 @@
...
+    uint16_t *ciphersuite;      /** < The lit of cipher suites */
+} DetectTlsCipherSuiteData;
+
+void DetectTlsCipherSuiteFree(void *data)
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: static?

File: src/detect-tls-cipher-suite.c [hunk 3]
```diff
@@ -0,0 +1,1093 @@
...
+ *
+ * \param value - 2 bytes value for cipher suite code
+ *
+ * \retval "-" if not found
+ * \retval const char * with description if found
+ */
+const char * TlsCipherSuiteDescription(uint16_t value)
+{
+#if __BYTE_ORDER == __BIG_ENDIAN
+    uint16_t I = value;
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: "I" name

File: src/detect-tls-cipher-suite.c [hunk 4]
```diff
@@ -0,0 +1,1093 @@
...
+ * \param t pointer to thread vars
+ * \param det_ctx pointer to the pattern matcher thread
+ * \param p pointer to the current packet
+ * \param m pointer to the sigmatch that we will cast into DetectTlsData
+ *
+ * \retval 0 no match
+ * \retval 1 match
+ */
+
+//
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: excess lines between comment block and func

File: src/detect-tls-cipher-suite.c [hunk 5]
```diff
@@ -0,0 +1,1093 @@
...
+        connp = &ssl_state->client_connp;
+    } else {
+        connp = &ssl_state->server_connp;
+    }
+
+    if (connp->num_cipher_suites > 0 && data->length > 0) {
+        for (uint32_t j=0; j < data->length; ++j) {
+            for (uint32_t i=0; i < connp->num_cipher_suites; ++i) {
+#if __BYTE_ORDER == __BIG_ENDIAN
+                uint16_t I = connp->cipher_suites[i];
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: I

File: src/detect-tls-cipher-suite.c [hunk 6]
```diff
@@ -0,0 +1,1093 @@
...
+    if (flags & STREAM_TOSERVER) {
+        connp = &ssl_state->client_connp;
+    } else {
+        connp = &ssl_state->server_connp;
+    }
+
+    if (connp->num_cipher_suites > 0 && data->length > 0) {
+        for (uint32_t j=0; j < data->length; ++j) {
+            for (uint32_t i=0; i < connp->num_cipher_suites; ++i) {
+#if __BYTE_ORDER == __BIG_ENDIAN
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: I would prefer having the byte order handling centralized somehow. Too many opportunities for mistakes this way.

File: src/detect-tls-cipher-suite.c [hunk 7]
```diff
@@ -0,0 +1,1093 @@
...
+                uint16_t I = connp->cipher_suites[i];
+#elif __BYTE_ORDER == __LITTLE_ENDIAN
+                uint16_t I = (connp->cipher_suites[i]>>8) | (connp->cipher_suites[i]<<8);
+#endif
+                uint16_t J = data->ciphersuite[j];
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: J

File: src/detect-tls-cipher-suite.c [hunk 8]
```diff
@@ -0,0 +1,1093 @@
...
+    sigmatch_table[DETECT_AL_TLS_CIPHER_SUITE].Free  = DetectTlsCipherSuiteFree;
+    sigmatch_table[DETECT_AL_TLS_CIPHER_SUITE].RegisterTests = DetectTlsCipherSuiteRegisterTests;
+
+    DetectSetupParseRegexes(PARSE_REGEX,
+            &cipher_parse_regex, &cipher_parse_regex_study);
+
+    g_tls_generic_list_id = DetectBufferTypeRegister("tls_generic");
+}
+
+static uint16_t TlsCipherSuiteNumLisEntries(const char *str) {
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: style: { on newline

File: src/detect-tls-cipher-suite.c [hunk 9]
```diff
@@ -0,0 +1,1093 @@
...
+    ret = pcre_exec(cipher_parse_regex, cipher_parse_regex_study,
+                    str, strlen(str), 0, 0,
+                    ov, MAX_SUBSTRINGS);
+
+    if (ret != 3) {
+        SCLogError(SC_ERR_PCRE_MATCH, "invalid cipher.suite option: %d", ret);
+        goto error;
+    }
+
+    res = pcre_get_substring((char *)str, ov, MAX_SUBSTRINGS, 1, &side_str);
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: please convert to pcre_copy_substring, makes memory handling easier

File: src/detect-tls-cipher-suite.c [hunk 10]
```diff
@@ -0,0 +1,1093 @@
...
+    if (res < 0) {
+        SCLogError(SC_ERR_PCRE_GET_SUBSTRING, "pcre_get_substring failed");
+        goto error;
+    }
+
+    res = pcre_get_substring((char *)str, ov, MAX_SUBSTRINGS, 2, &list_str);
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: pcre_copy_substring

File: src/detect-tls-cipher-suite.c [hunk 11]
```diff
@@ -0,0 +1,1093 @@
...
+    for (size_t i=0; i < len; ++i) {
+        if (*ptr == ':') {
+            ++ret;
+        }
+        ++ptr;
+    }
+    return ret;
+}
+
+static DetectTlsCipherSuiteData *DetectTlsCipherSuiteParse (char *str)
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: const char *str

File: src/detect-tls-cipher-suite.c [hunk 12]
```diff
@@ -0,0 +1,1093 @@
...
+
+    res = pcre_get_substring((char *)str, ov, MAX_SUBSTRINGS, 2, &list_str);
+    if (res < 0) {
+        SCLogError(SC_ERR_PCRE_GET_SUBSTRING, "pcre_get_substring failed");
+        goto error;
+    }
+
+    uint32_t direction = 0 ;
+
+    if (SCMemcmp("client", side_str, 6) == 0) {
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: side_str is a string, so use string functions like strcmp

File: src/detect-tls-cipher-suite.c [hunk 13]
```diff
@@ -0,0 +1,1093 @@
...
+                uint16_t value = (int)strtol(pch, NULL, 16);
+                data->ciphersuite[rule_suite_index] = value;
+                ++rule_suite_index;
+                break;
+            }
+        }
+
+        /* Find by description */
+        for (size_t i=0; i < sizeof(g_suites); ++i) {
+            if ( SCMemcmp(pch, g_suites[i].description, strlen(g_suites[i].description)) == 0 ) {
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: strcmp

File: src/detect-tls-cipher-suite.c [hunk 14]
```diff
@@ -0,0 +1,1093 @@
...
+    if (data == NULL)
+        goto error;
+
+    /* Okay so far so good, lets get this into a SigMatch
+     * and put it in the Signature. */
+    sm = SigMatchAlloc();
+    if (sm == NULL)
+        goto error;
+
+    if (s->alproto != ALPROTO_UNKNOWN && s->alproto != ALPROTO_TLS) {
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: Replace with DetectSignatureSetAppProto like in other tls keywords.

File: src/detect-tls-cipher-suite.c [hunk 15]
```diff
@@ -0,0 +1,1093 @@
...
+    de_ctx = DetectEngineCtxInit();
+    FAIL_IF_NULL(de_ctx);
+
+    de_ctx->flags |= DE_QUIET;
+    de_ctx->sig_list = SigInit(de_ctx, "alert tls any any -> any any "
+                               "(msg:\"Testing client tls_cipher_suite\"; "
+                               "tls_cipher_suite:client:0x0000; sid:1;)");
+    FAIL_IF_NULL(de_ctx->sig_list);
+
+    SigGroupCleanup(de_ctx);
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: just DetectEngineCtxFree is enough (lots of tests still have a bad pattern)

File: src/detect-tls-cipher-suite.c [hunk 16]
```diff
@@ -0,0 +1,1093 @@
...
+    r = AppLayerParserParse(NULL, alp_tctx, &f, ALPROTO_TLS, STREAM_TOCLIENT,
+                            server_hello, sizeof(server_hello));
+    FLOWLOCK_UNLOCK(&f);
+
+    FAIL_IF(r != 0);
+
+    SigMatchSignatures(&tv, de_ctx, det_ctx, p2);
+    FAIL_IF_NOT(PacketAlertCheck(p2, 2) == 1);
+
+    if (alp_tctx != NULL)
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: these conditions can go away, it should be checked with FAIL_IF_NULL instead above, right after the pointers are set

File: src/detect-tls-cipher-suite.c [hunk 17]
```diff
@@ -0,0 +1,1093 @@
...
+}
+
+/**
+ * \brief Registration function for keyword: tls_cipher_suite
+ */
+void DetectTlsCipherSuiteRegister(void)
+{
+    sigmatch_table[DETECT_AL_TLS_CIPHER_SUITE].name = "tls_cipher_suite";
+    sigmatch_table[DETECT_AL_TLS_CIPHER_SUITE].desc = "match SSLv2, SSLv3, TLSv1, TLSv1.1 and TLSv1.2 Cipher Suites";
+    sigmatch_table[DETECT_AL_TLS_CIPHER_SUITE].url = DOC_URL DOC_VERSION "/rules/tls-keywords.html#cipher.suite";
```
Comment (inliniac) [inline src/detect-tls-cipher-suite.c]: is that url correct? The dot looks suspicious

**Comment by inliniac** [comment]
Please make changes in a new branch followed by a new PR, thanks!

File: src/app-layer-ssl.c [hunk 18]
```diff
@@ -80,6 +80,340 @@ SCEnumCharMap tls_decoder_event_table[ ] = {
     { NULL,                          -1 },
 };
 
+// http://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-4
+static SSLCipherSuite g_suites[] = {
```
Comment (jasonish) [inline src/app-layer-ssl.c]: But also keep a fallback in the code?

Comment (inliniac) [inline src/app-layer-ssl.c]: Probably a good idea, yes.

**Comment by fooinha** [comment]
Replaced by #2657

**Pull request closed without merging**